tty: set the handle to blocking mode

Fishrock123 opened the pull request:
<!--
Thank you for your pull request. Please review below requirements and walk
through the checklist. You can 'tick' a box by using the letter "x": [x].

Run the test suite with: `make -j4 test` on UNIX or `vcbuild test nosign` on
Windows.

If this aims to fix a regression or you’re adding a feature, make sure you also
write a test. If possible, include a benchmark that quantifies your changes.

Finally, read through our contributors guide and make adjustments as necessary:
https://github.com/nodejs/node/blob/master/CONTRIBUTING.md
-->

##### Checklist
<!-- remove lines that do not apply to you -->

- [X] tests and code linting passes
- [ ] a test and/or benchmark is included
- [x] documentation is changed or added
- [X] the commit message follows commit guidelines


##### Affected core subsystem(s)
<!-- provide affected core subsystem(s) (like doc, cluster, crypto, etc) -->
tty, process

##### Description of change
<!-- provide a description of the change below this comment -->

See discussion at:
- https://github.com/nodejs/node/pull/177#issuecomment-119351671
- https://github.com/nodejs/node/issues/6456#issuecomment-219620603
- https://github.com/nodejs/node/pull/6773

cc @nodejs/ctc @isaacs @ksc @saghul etc

Note: we still have more investigation to do but this is looking like it is probably the correct thing to do.

indutny commented:
I have a deja vu here... Is it like 3rd time we re-enable it?

indutny commented:
Am I right that running `node script.js | some-prog-that-does-not-read-stdin` will make `script.js` eventually hang on writes?

addaleax commented:
@indutny This should only affect `tty`s, so, no, I don’t think so.

indutny commented:
@addaleax oh, right! Thank you. LGTM then, I have no issues with this.

mscdex commented:
Shouldn't there at least be an official way to change the behavior (in case someone needs/wants async)?

Fishrock123 commented:
> I have a deja vu here... Is it like 3rd time we re-enable it?

Haha, not quite. ;)

@mscdex Official? Idk. But you can still do `tty._handle.setBlocking(false)` afaik.

That being said, I can't think of a reason you'd need async in a TTY? Maybe if only one stream is to a TTY?

mscdex commented:
@Fishrock123 I meant a way without accessing an underscored property.

saghul commented:
LGTM. (fingers crossed so we don't open a can of worms).

/cc @bnoordhuis I believe this whole thing make some sense in your brain, I'd love to see your yay or nay here.

jasnell commented:
@mscdex ... like this? https://github.com/jasnell/node/commit/4982550d84b760d01a75739a584596101d3844d3

jasnell commented:
This PR should include a documentation update that (a) describes the default behavior of `tty.ReadStream` and (b) demonstrates a supported way of making it unblocking. I'm not comfortable with this landing until at least (a) is added.

Fishrock123 commented:
@jasnell So, there are a couple issues with your API proposal.

The biggest is that actually creating TTY streams isn't particularly supported:

> In normal circumstances, `process.stdout` will be the only `tty.WriteStream` instance ever created (and only when `isatty(1)` is `true`).

In fact, I've tried to make tty streams before and found it's virtually impossible. You can only make one if the `fd` is a tty and it is not already in use, which is basically never. I actually don't know of a case when you could ever make one manually without severe hackery.

However, exposing a method publicly also causes some concerns:

1. The existing problem that `stdio` streams do not have consistent APIs depending on the stream type is yet again amplified (very difficult to fix)
2. It doesn't work for pipes, but putting this functionality into `net.Socket` publicly is harmful because of TCP connections.
3. I'm not sure unblocking is ever valuable for this, in any manor that isn't strictly theoretical.
  - Same goes for pipes, where blocking only has theoretical niceties.

--

> describes the default behavior of `tty.ReadStream`

I doubt that is useful due to the above, or the correct place to put it, since no-one will go looking for the TTY module from `console.log()` or `process.std{in|out|err}`.

I will add a note to where the existing note is for file piping though: https://nodejs.org/dist/latest-v6.x/docs/api/process.html#process_process_stderr

Fishrock123 commented:
Updated with docs note.

addaleax commented on doc/api/process.md:
```diff
@@ -1353,11 +1353,15 @@ if (process.getuid && process.setuid) {
+  - (Although disks are fast and operating systems normally employ write-back
+    caching so this is very uncommon.)
+4. Writes __will__ block by default if output is going to a TTY (a terminal) or
+Node.js is running on windows.
```
nit: It’s usually stylized as `Windows` in the docs.

addaleax commented:
LGTM

mscdex commented:
@jasnell Something like that would be ideal of course, but even if there was a way to simply change the behavior once on startup. I'm not sure what that would look like or if enough people would be interested in supporting something like that though ...

jasnell commented:
@mscdex ... I'm definitely for it (either API or command line switch) but given the discussion I doubt we'd get consensus to add it so I'm going to just drop it.

jasnell commented:
I would say that this looks like a semver-major change but I know there will be disagreement with that.

addaleax commented:
@jasnell Could you give an example of what you imagine might break because of do this change?

Fishrock123 commented:
I can imagine a lot of things that are already breaking _without_ this change. ;)

Fishrock123 commented:
@jasnell @mscdex I'd be open to it if there is actually a reason to do it, but no reasons are appearing...?

jasnell commented:
@addaleax ... not off hand but it is a change in the established default behavior, which would technically qualify it as a semver-major. Note that I didn't add the label ;-) ... I'm quite certain that I'm in the minority.

Fishrock123 commented:
> but it is a change in the established default behavior,

Sorta. It gets tricky here because on OS X it appears to have previously actually blocked when it wasn't supposed to so the majority of the CLI users never noticed this because it never changed on their platform sooner. So... the previous behavior is actually more established there, and due to previous versions of node.js, also across the board I think. Additionally, I'm unsure if `console.log()` blocks in the browser but I would not be particularly surprised if it did.

mscdex commented:
I'm more inclined to say `semver-major` too since the non-blocking behavior in general has existed for quite a long time now in node so I don't view this change as a "oops we introduced a bug, let's fix it quickly" kind of situation. \*shrug\*

addaleax commented:
As far as I can tell, the only visible difference would be in timing, no? And unlike with pipes, for `tty`s there is (in 99.99 % of practically occurring situations) an active consumer on the other end, so writing won’t block indefinitely because of this.

addaleax commented:
(If people want to hold this off until v7, sure, do that. All I’m trying to say is that I don’t believe there’s any reason that’s necessary.)

Fishrock123 commented:
> I'm more inclined to say `semver-major` too since the non-blocking behavior in general has existed for quite a long time now in node

Perhaps, but consider that `console.log()` doesn't even look async; there's no callback. Also, we block on windows and it doesn't create any issues there.

It seems like blocking is done only for performance and pipe reasons.

This doesn't effect the latter, and TTYs are fast so the former does not seem to be an issue.

We are going from slightly less deterministic to fully deterministic. I still can't see a problem...

--

If everyone feels more comfortable, I can make this the case only on OS X, since it was previously always blocking on OS X, but I doubt that is a good solution.

trevnorris commented:
stdio has always blocked on Windows. So in this case we've had a documentation error. Also leading to UB between platforms. This is a bug fix. Not a feature being changed.

saghul commented:
Please, hold a bit, I think we got it: https://github.com/nodejs/node/issues/6456#issuecomment-219974514

Fishrock123 commented:
Nice to get confirmation, looks like the solution on our end is the same still.

saghul commented:
Yep, sorry, carry on with this fix. It was a false alarm triggered by another bug. 😢

orangemocha commented:
Related: https://github.com/nodejs/node-v0.x-archive/issues/3584

Fishrock123 commented:
cc @piscisaureus @bnoordhuis Could you *please* bring any extra background and concerns to this?

Also, I will be adding an environment variable shortly for telling Node.js to use non-blocking if possible, to allow linux to non-block to TTYs if necessary, or for remote TTYs.

Fishrock123 commented:
Add the `NODE_ASYNC_STDIO` env variable checking and related documentation (mostly).

saghul commented:
Since the env variable only applies to ttys, shouldn't it include "TTY" in the name?

Fishrock123 commented:
@saghul Perhaps, but I was thinking this was the most generic if we needed to expand it in the future. Perhaps it would be `NODE_ASYNC_STDIO=<subsystem>` then? Idk it's hypothetical.

piscisaureus commented:
Okay, couple of things:

* TTY != stdio. Stdio *can* refer to a tty, if the output isn't redirected to a file or a pipe.
* Setting the blocking mode on windows is only supported for pipes - not for console output.
* Pipe stdio was traditionally non-blocking on windows; however on unix systems it was blocking. I'm not quite sure what the current situation is. That's the exact opposite of what @trevnorris said.
* The reason stdio was made blocking is that people complained about output being clipped if node crashes. It's pretty difficult to debug an application if the error doesn't make it to the console/logfile. 

Now, I know this is a confusing mess.

My suggestion would be to find a way to implement `writeSync` for streams that are otherwise non-blocking. If you do that you can just make stdio streams nonblocking like everything else. `console.log` and `console.error` could use this under the hood, whilst interaction through the streams API would be asynchronous as people expect.

An environment variable is a good stopgap of course if you want to get home quick. But please please make it a temporary thing. I predict it'll cause a steady influx of bug reports to node and the ecosystem. First candidates "out of memory when using cluster under load" and "no output in webstorm" and "forever hangs when stdio is nonblocking". 
Not to mention https://github.com/nodejs/node-v0.x-archive/issues/3871 and https://github.com/nodejs/node-v0.x-archive/issues/6249.
Good luck folks! :)

Fishrock123 commented:
> * TTY != stdio. Stdio *can* refer to a tty, if the output isn't redirected to a file or a pipe.

Right, this is about making stdio blocking to TTYs. I can rename the env variable.

> * Setting the blocking mode on windows is only supported for pipes - not for console output.

Oh. As in, non-blocking does not work on windows? That throws a curveball into this. cc @nodejs/platform-windows?

> * Pipe stdio was traditionally non-blocking on windows; however on unix systems it was blocking. I'm not quite sure what the current situation is. That's the exact opposite of what @trevnorris said.

I'm not sure about windows. On unix systems it is now non-blocking as of v1.0.2 (libuv 1.2.1?).

> The reason stdio was made blocking is that people complained about output being clipped if node crashes. It's pretty difficult to debug an application if the error doesn't make it to the console/logfile. 

well doesn't that sound familiar...

> I predict it'll cause a steady influx of bug reports to node and the ecosystem. First candidates "out of memory when using cluster under load" and "no output in webstorm" and "forever hangs when stdio is nonblocking". 
Not to mention https://github.com/nodejs/node-v0.x-archive/issues/3871 and https://github.com/nodejs/node-v0.x-archive/issues/6249.

You are describing the current state of things.

Except that linux has a large enough buffer size that this typically becomes a problem on other platforms first..

Fishrock123 commented:
I've made https://github.com/nodejs/node/pull/6895 as a temporary fix for the OS X problem in specific while we investigate other platforms. We should get it in for next weeks v6 release and continue to look into how to fix the overarching problem.

kzc commented:
Another data point to be aware of... Terminal applications vary wildly in tty write speed so please don't assume Linux blocking tty writes are slow just based on `gnome-terminal` performance on Ubuntu (for example).

On my Linux VM it takes `gnome-terminal` 3.9s of wall clock time to do an `ls -lR .` in the node source root directory. `urxvt` on the other hand just takes 0.6s for the same operation - a 6.5X speed difference.

Another informal benchmark of terminal application print speed I found from 2010:
```
wterm           0.3s
aterm           0.3s
rxvt            0.3s
mrxvt           0.4s
konsole         0.6s
yakuake         0.7s
lxterminal        7s
xterm             9s
gnome-terminal   12s
xfce4-terminal   12s
vala-terminal    18s
xvt              48s
```
http://stackoverflow.com/a/3860319

Relative performance may have changed since then, but you get the idea.

jasnell commented:
Does this one need to remain open?

kzc commented:
This issue was never resolved. stdio tty writes over 64K at the console will still be truncated on Linux upon process.exit().

Only Mac having blocking stdio is a strange design choice for a portable programming language.

jasnell commented:
@kzc the issue is still unresolved, yes. I was referring to this specific PR tho.

kzc commented:
I was just highlighting the inconsistency of Mac having a differing behavior from all other UNIX variants. Either they all should block stdio writes, or none should. Mac blocking stdio is just masking a basic design problem.

saghul commented:
I agree with @kzc, but at this point I'm not sure what keeping this PR open gives us.

Fishrock123 commented:
I still don't see why we wouldn't merge a version of this (perhaps with a better named ENV variable, or none at all).

I don;t have access to a linux machine for a while, so I can't really test TTY speeds. If someone else wants to test if non-blocking is significantly faster, feel free.

kzc commented:
tty speed is entirely dependent on the quality of the terminal program: https://github.com/nodejs/node/pull/6816#issuecomment-221646278

Fishrock123 commented:
@kzc Did you try testing those numbers of non-blocking vs blocking? Is there any speed difference in the end?

kzc commented:
Those would be blocking tty numbers of single-threaded non node programs. It shows the throughput limit of each terminal application. There's no need to test non-blocking numbers as in the case of node on Linux as it would be truncating data over 64K in the event of process.exit() so it wouldn't be a fair comparison. 

Correctly outputting all stdio data upon node exiting by uncaught exception or process.exit() should be the primary consideration. If that can be achieved without blocking stdio writes that would be ideal. Blocking stdio is just a stop-gap measure until a better solution is found. In any case, blocking tty writes does not address truncation of output when stdout/stderr is piped. It is evident that the stdio subsystem of node needs a redesign.

Fishrock123 commented:
> Correctly outputting all stdio data upon node exiting by uncaught exception or process.exit() should be the primary consideration.

@kzc Unfortunately, that's not exactly how this works. I need to convince the rest of the collaborators that there won't be perf troubles here so that it could be landed as a patch fix.

... I guess this could be landed as a major but IMO that would be silly, should be a patch.

kzc commented:
> Unfortunately, that's not exactly how this works. I need to convince the rest of the collaborators that there won't be perf troubles here so that it could be landed as a patch fix.

I'm not a fan of having any node platform block stdio - I'm just arguing for consistency and portability. If Mac has this behavior then all platforms should.

In the bigger picture Node ought to guarantee flushing stdout and stderr (whether to console or piped) upon exit or uncaught exception just like every other programming language out there. It's hard to fathom that a case has to be made for this. 

If producing the wrong output quickly is the primary consideration of NodeJS then all stdio truncating bugs may as well be closed as WONTFIX. I give up. I don't have any more time to devote to this project.

Fishrock123 commented:
@nodejs/ctc Can we please do this? All evidence points to it not being a problem, and it is a terrible behavior inconsistency with potential problems without this.

I have updated and renamed the ENV variable.

rvagg commented on doc/api/cli.md:
```diff
@@ -288,6 +288,17 @@ Path to the file used to store the persistent REPL history. The default path is
+-->
+
+When set to `1` writes stdout and stderr will be non-blocking and async when
+outputting to a TTY on platforms which support async stdio.
```
Can we expand on this with specifics? Do we have an accurate understanding of Windows now (I can't keep track of how many times both sides have been asserted for Windows)?

Fishrock123 replied:
> Do we have an accurate understanding of Windows

Erm, maybe.

I'm 99% sure Windows is always async to a Terminal because windows doesn't give you a way to tell if a write ever actually made it to the terminal. (Sources: Bert and Trevor, I think.)

Pipes definitely block on windows. (Source: [this code in `net.js`](https://github.com/nodejs/node/blob/c09dfe3a0bf63ecff522a7f9c06c21593969e69d/lib/net.js#L140-L144))

I have no idea about files & windows.

Fishrock123 replied:
Note that `setBlocking()` does nothing if the platform doesn't support it.

rvagg commented on doc/api/cli.md:
```diff
@@ -288,6 +288,17 @@ Path to the file used to store the persistent REPL history. The default path is
+When set to `1` writes stdout and stderr will be non-blocking and async when
+outputting to a TTY on platforms which support async stdio.
+Setting this will void any guarantee that stdio will not be interleaved or
+dropped at program exit. **Avoid use.**
```
perhaps soften, something like `**Use of this mode is not recommended.** The name "UNSAFE" already is already blunt enough I reckon.

rvagg commented on doc/api/process.md:
```diff
@@ -1429,15 +1428,20 @@ Android)
-emit the [`'finish'`][] event, and writes can block when output is redirected to
-a file (although disks are fast and operating systems normally employ write-back
-caching so it should be a very rare occurrence indeed.)
+Note: `process.stderr` and `process.stdout` are differ from other streams in
```
remove "are"

rvagg commented on doc/api/process.md:
```diff
@@ -1429,15 +1428,20 @@ Android)
+1. They cannot be closed ([`end()`][] will throw).
+2. They never emit the [`'finish'`][] event.
+3. Writes _can_ block when output is redirected to a file.
+  - (Although disks are fast and operating systems normally employ write-back
```
this doesn't make sense bracketed as a lone sub-point, remove sub-point and bring it back up to the previous line, I'd even remove the brackets but that up to you.

rvagg commented on doc/api/process.md:
```diff
@@ -1429,15 +1428,20 @@ Android)
+    caching so this is very uncommon.)
+4. Writes on UNIX __will__ block by default if output is going to a TTY
+   (a terminal).
+5. Windows functionality differs. Writes block except when output is going to a
```
isn't this PR reversing that for Windows though? "Writes block whether writing to a TTY or not."?

Fishrock123 replied:
That should not be possible on windows, iirc  `setBlocking()` does nothing on that platform (or any platform) where it is not supported.

Fishrock123 replied:
ok maybe cc @saghul on this, since I don't really know where to look in libuv besides `uv_stream_set_blocking`. It appears to set `UV_HANDLE_BLOCKING_WRITES` on Windows but I don't know how to for sure tell if it only affects pipes.

A search only shows results in `win/pipe.c` and `win/stream.c`.

rvagg commented on doc/api/process.md:
```diff
@@ -1482,11 +1486,17 @@ console.log = (msg) => {
-emit the [`'finish'`][] event and that writes can block when output is
-redirected to a file (although disks are fast and operating systems normally
-employ write-back caching so it should be a very rare occurrence indeed.)
+Note: `process.stderr` and `process.stdout` are differ from other streams in
```
can we remove this duplication by pulling it up as a top-level topic and link to it from these sections?

rvagg commented:
YES, LGTM sans minor doc nits. Let's get this merged and ship it, we've been in an awkward limbo on this for too long. Any disagreement from @nodejs/collaborators at this stage or have we exhausted ourselves on this?

In the longer term I really like @piscisaureus' suggestion:

> My suggestion would be to find a way to implement `writeSync` for streams that are otherwise non-blocking. If you do that you can just make stdio streams nonblocking like everything else. `console.log` and `console.error` could use writeSync under the hood, whilst interaction through the streams API would be asynchronous as people expect.

But I honestly have zero clue about the feasability of that across supported platforms. Can anyone suggest a possible path forward on that?

targos commented on doc/api/cli.md:
```diff
@@ -288,6 +288,17 @@ Path to the file used to store the persistent REPL history. The default path is
+added: REPLACEME
+-->
+
+When set to `1` writes stdout and stderr will be non-blocking and async when
```
writes **to** stdout and stderr ?

Fishrock123 commented:
I _think_ I addressed all the comments.

targos commented:
LGTM

Trott commented on doc/api/cli.md:
```diff
@@ -288,6 +288,17 @@ Path to the file used to store the persistent REPL history. The default path is
+added: REPLACEME
+-->
+
+When set to `1` writes to stdout and stderr will be non-blocking and async when
```
Nit: comma after `1`

Nit: backticks around `stdout` and `stderr` 

Nit: `async`->`asynchronous`

Trott commented on doc/api/cli.md:
```diff
@@ -288,6 +288,17 @@ Path to the file used to store the persistent REPL history. The default path is
+-->
+
+When set to `1` writes to stdout and stderr will be non-blocking and async when
+outputting to a TTY on platforms which support async stdio.
```
Nit: replace `outputting` with `writing` or some appropriate verb. (Using `output` as a verb rather than a noun is awkward, at least to me.)

Nit: I won't leave these specific nits everywhere but `async` is better (in my opinion, anyway) as `asynchronous` and all the stdio/stdin/stdout/stderr abbreviations should either be in backticks or else written out rather than shortened. (My opinion only, nits, you can ignore them, etc.)

Fishrock123 replied:
stdio is not a code thing, so I'm not going to backtick it.

Also people really should not use this anyways, so I'm not sure it matters if it is advanced wording.

Trott commented on doc/api/process.md:
```diff
@@ -1429,15 +1428,20 @@ Android)
-emit the [`'finish'`][] event, and writes can block when output is redirected to
-a file (although disks are fast and operating systems normally employ write-back
-caching so it should be a very rare occurrence indeed.)
+Note: `process.stderr` and `process.stdout` differ from other Node.js streams
```
@jasnell I believe you are trying to standardize the format for `Note:` stuff in the docs, right? What did you settle on?

jasnell replied:
`*Note*:` would be the preference. Once I'm done going through all of the docs and standardizing the syntax I plan to work changes to the actual rendering of the notes in the docs based on the common format

Trott commented on doc/api/process.md:
```diff
@@ -1429,15 +1428,20 @@ Android)
-a file (although disks are fast and operating systems normally employ write-back
-caching so it should be a very rare occurrence indeed.)
+Note: `process.stderr` and `process.stdout` differ from other Node.js streams
+in a couple ways:
```
`in a couple ways:` -> `in several ways:` or `in a few ways:` or `in that:` ("a couple" literally means 2)

Trott commented on doc/api/process.md:
```diff
@@ -1429,15 +1428,20 @@ Android)
+2. They never emit the [`'finish'`][] event.
+3. Writes _can_ block when output is redirected to a file.
+  - Note that disks are fast and operating systems normally employ write-back
+    caching so this is very uncommon.
```
Nit: remove `very`

Fishrock123 replied:
Not comfortable with that, that was the original wording.

Fishrock123 replied:
@thefourtheye Pardon? Could you elaborate?

Trott replied:
Yes, there should be a comma before `so`. 

Also: `very uncommon` -> `uncommon`

Honestly, I would consider dispensing of the explanation entirely and just say: `This is uncommon.`

Trott commented on doc/api/process.md:
```diff
@@ -1482,11 +1486,17 @@ console.log = (msg) => {
-redirected to a file (although disks are fast and operating systems normally
-employ write-back caching so it should be a very rare occurrence indeed.)
+Note: `process.stderr` and `process.stdout` differ from other Node.js streams
+in a couple ways:
```
Same nits as above.

Trott commented on doc/node.1:
```diff
@@ -169,6 +169,13 @@ Path to the file used to store the persistent REPL history. The default path
 
+.TP
+.BR NODE_TTY_UNSAFE_ASYNC=1
+When set to 1 writes to stdout and stderr will be non-blocking and async when
```
Similar nits as above similar text.

Trott commented:
Left some nits on the documentation prose, but nothing that I think **must** happen before landing. (In other words: Don't interpret my nits as me objecting to this landing.)

TheAlphaNerd commented:
@Fishrock123 is this something that should be included in v4.5.0?

saghul commented:
@kzc A note to remember: when stdio was made blocking on macOS that was meant to restore the behavior aas in v4.x On other Unix systems it never blocked. We are changing that now. Is it a good or a bad thing? I no longer know.

kzc commented:
@saghul I know the history well. The blocking on Mac was initially inadvertently introduced as result of a bug pre-node 4.x when trying to re-open the tty - ironically to not block stdio - and somehow this blocking stdio behavior on Mac became grandfathered as the status quo. 

Whether this PR is good or bad, it just addresses that inconsistency on UNIX platforms. https://github.com/nodejs/node/pull/6816#issuecomment-231762130 sums up my view. Blocking stdio tty alone is not enough to fix the many long standing problems with node stdio. Something like https://github.com/nodejs/node/pull/6773 is still needed to resolve truncated piped stdout/stderr and some similar fix is still required to flush piped stdio in the event of an uncaught exception. Once node is in the process of an abrupt exit and the event loop is no longer running (although not destroyed) there's zero downside to flushing stdout and stderr as users would reasonably expect.

As I don't watch these node issue/PR threads any longer please my @ handle if you want to get any other feedback.

saghul commented:
@kzc Sure, I know we are currently not in a very good place. Thanks for your efforts so far.

Fishrock123 commented:
@TheAlphaNerd Give it's potentially debatable if this is a major, I'm going to say no.

Fishrock123 commented:
I consider it a non-breaking bugfix given that:
- There are no shortage of complaints about the current behavior
- There is no evidence this could cause problems

Fishrock123 commented:
New CI: https://ci.nodejs.org/job/node-test-pull-request/3324/

@jasnell Could you take a look at this again and let us know if you are comfortable with it?

thefourtheye commented on doc/api/process.md:
```diff
@@ -1429,15 +1428,20 @@ Android)
+Note: `process.stderr` and `process.stdout` differ from other Node.js streams
+in several ways:
+1. They cannot be closed ([`end()`][] will throw).
+2. They never emit the [`'finish'`][] event.
```
We don't need the additional square brackets for links I guess.

Fishrock123 replied:
Huh? These should use links defined at the bottom of the file...?

thefourtheye replied:
If we have defined them at the bottom the empty square brackets are not necessary I think

thefourtheye commented on doc/api/process.md:
```diff
@@ -1482,11 +1486,17 @@ console.log = (msg) => {
+2. They never emit the [`'finish'`][] event.
+3. Writes _can_ block when output is redirected to a file.
+  - Note that disks are fast and operating systems normally employ write-back
+    caching so this is very uncommon.
```
There are two sentences here

Fishrock123 replied:
@thefourtheye Pardon? Could you elaborate?

thefourtheye replied:
Shouldn't there be a comma or semicolon before so?

jasnell commented on lib/tty.js:
```diff
@@ -48,12 +48,12 @@ function WriteStream(fd) {
   // even though it was originally intended to change in v1.0.2 (Libuv 1.2.1).
   // Ref: https://github.com/nodejs/node/pull/1771#issuecomment-119351671
-  if (process.platform === 'darwin') this._handle.setBlocking(true);
+  this._handle.setBlocking(process.env.NODE_TTY_UNSAFE_ASYNC !== '1');
```
Setting a `const` at module load would be microscopically faster :-)

jasnell commented:
LGTM but we'll need to message this properly. While I believe there is consensus around landing this as a bug fix, it is technically a fairly significant change in behavior that *should* technically be tagged as a semver-major. There should be a release candidate cycle for any release that this goes out in just to be on the safe side.

Fishrock123 commented:
If it must land in a semver-major, it could.

I do not think it is reasonable to see this as an API change however, so I think it may be worth our while to get this in before v6 goes LTS.

addaleax commented:
This still LGTM, and I’m still +1 on not seeing this as a semver-major because afaict, this doesn’t necessarily represent a visible behavioural change.

Fishrock123 commented:
Guess we're clear to land... or as clear as it will get.

@jasnell how do you propose we message it?

jasnell commented:
Twitter PSA at the very least (I can do that in just a minute). Even tho this will land as a patch, we need to make sure it's called out explicitly in the release notes. Beyond that, I'm not sure what else.

jasnell commented:
I would definitely recommend that we do a release candidate for whatever release this lands in.

addaleax commented:
I’ll be landing this later today if there are no objections and nobody else does before me, this has been lying around long enough.

addaleax commented:
Landed in ab3306ad51d8136014aa0fa9278b57bb77105320
